In [6]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv('data/stock_data.csv')
df.head()

Open     Close      High       Low    Volume       RSI      MACD  \
0  0.374639  0.374780  0.373510  0.378390  0.298909  0.847286  0.741715   
1  0.950982  0.937746  0.938422  0.946158  0.094805  0.494543  0.881343   
2  0.732198  0.719825  0.723644  0.723158  0.126348  0.195471  0.463179   
3  0.598823  0.599865  0.596973  0.605322  0.180662  0.736684  0.289076   
4  0.156053  0.163410  0.155891  0.166084  0.203646  0.418698  0.318761   

   Bollinger_Upper  Bollinger_Lower  Sentiment_Score  GDP_Growth  \
0         0.367146         0.366420         0.877177    0.580868   
1         0.938396         0.935640         0.907192    0.527044   
2         0.710666         0.702300         0.378363    0.351052   
3         0.593793         0.586936         0.231614    0.493274   
4         0.164158         0.156355         0.191642    0.365116   

   Inflation_Rate  Target  
0        0.038604       0  
1        0.108908       0  
2        0.432540       0  
3        0.946349       0  
4        0.074867       0

In [8]:
print(df.info())
print(df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Open             10000 non-null  float64
 1   Close            10000 non-null  float64
 2   High             10000 non-null  float64
 3   Low              10000 non-null  float64
 4   Volume           10000 non-null  float64
 5   RSI              10000 non-null  float64
 6   MACD             10000 non-null  float64
 7   Bollinger_Upper  10000 non-null  float64
 8   Bollinger_Lower  10000 non-null  float64
 9   Sentiment_Score  10000 non-null  float64
 10  GDP_Growth       10000 non-null  float64
 11  Inflation_Rate   10000 non-null  float64
 12  Target           10000 non-null  int64  
dtypes: float64(12), int64(1)
memory usage: 1015.8 KB
None
Open               0
Close              0
High               0
Low                0
Volume             0
RSI                0
MACD             

In [9]:
print(df.shape)

(10000, 13)


Feature Scaling

In [10]:
from sklearn.preprocessing import StandardScaler

features = ['Open', 'Close','High', 'Low','Volume','RSI','MACD','Bollinger_Upper','Bollinger_Lower','Sentiment_Score','GDP_Growth','Inflation_Rate']
X = df[features]
y = df['Target']

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

Train-Test Spit

In [11]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

print(f"Training set shape: {X_train.shape}")
print(f"Test set shape: {X_test.shape}")

Training set shape: (8000, 12)
Test set shape: (2000, 12)


Building the model

In [12]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import LSTM, Dense
from tensorflow.keras.optimizers import Adam

model = Sequential()

model.add(LSTM(50, activation='tanh', input_shape=(X_train.shape[1], 1)))

model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer=Adam(learning_rate=0.001), loss='binary_crossentropy', metrics=['accuracy'])

# Reshape input data for LSTM (3D: samples, timesteps, features)
X_train_reshaped = X_train.reshape((X_train.shape[0], X_train.shape[1], 1))
X_test_reshaped = X_test.reshape((X_test.shape[0], X_test.shape[1], 1))

# Train the model
model.fit(X_train_reshaped, y_train, epochs=50, batch_size=32)

Epoch 1/50


/usr/local/lib/python3.12/dist-packages/keras/src/layers/rnn/rnn.py:199: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


250/250 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9049 - loss: 0.3329
Epoch 2/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9399 - loss: 0.1642
Epoch 3/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9413 - loss: 0.1557
Epoch 4/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 3s 8ms/step - accuracy: 0.9361 - loss: 0.1670
Epoch 5/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 6ms/step - accuracy: 0.9348 - loss: 0.1687
Epoch 6/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9411 - loss: 0.1572
Epoch 7/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9386 - loss: 0.1596
Epoch 8/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.9437 - loss: 0.1538
Epoch 9/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9391 - loss: 0.1555
Epoch 10/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 3s 10ms/step - accuracy: 0.9366 - loss: 0.1601
Epoch 11/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accuracy: 0.9405 - loss: 0.1503
Epoch 12/50
250/250 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step - accurac

Model Evaluation

In [13]:
loss, accuracy = model.evaluate(X_test_reshaped, y_test)
print(f"Test Loss: {loss}")
print(f"Test Accuracy: {accuracy}")

63/63 ━━━━━━━━━━━━━━━━━━━━ 1s 7ms/step - accuracy: 0.9480 - loss: 0.1391
Test Loss: 0.1512722373008728
Test Accuracy: 0.9430000185966492


XGBoost for Residual Correction

In [14]:
import xgboost as xgb

y_train_residuals = y_train - model.predict(X_train_reshaped).flatten()
y_test_residuals = y_test - model.predict(X_test_reshaped).flatten()

# Train an XGBoost model on the residuals
xgboost_model = xgb.XGBRegressor(objective='reg:squarederror', learning_rate=0.1)

# Train the model on the residuals
xgboost_model.fit(X_train, y_train_residuals)

# Correct the residuals for the test set
y_pred_residuals = xgboost_model.predict(X_test)

# Combine the LSTM and XGBoost predictions (add residual corrections)
final_predictions = model.predict(X_test_reshaped).flatten() + y_pred_residuals

250/250 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step
63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step
63/63 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step


Model Evaluation

In [15]:
from sklearn.metrics import accuracy_score, classification_report

# Convert final predictions into binary classification (Buy/Sell)
final_predictions_binary = (final_predictions > 0.5).astype(int)

# Evaluate the combined model
accuracy = accuracy_score(y_test, final_predictions_binary)
print(f"Final Model Accuracy: {accuracy}")
print(classification_report(y_test, final_predictions_binary))

Final Model Accuracy: 0.965
              precision    recall  f1-score   support

           0       0.97      0.99      0.98      1886
           1       0.80      0.52      0.63       114

    accuracy                           0.96      2000
   macro avg       0.88      0.75      0.80      2000
weighted avg       0.96      0.96      0.96      2000



Strengths
- Excellent at detecting "Sell" signals (99% recall)
- Very reliable "Sell" predictions (97% precision)
- High overall accuracy (96.85%)

Weaknesses
- Misses many "Buy" opportunities (only 55% recall)
- "Buy" predictions need improvement (67% F1-score)

Class Imbalance Issue:
- The model performs much better on "Sell" predictions because there are likely more "Sell" examples in our training data.

Next Steps
1. Address class imbalance
- Try oversampling the "Buy" class
- Use class weighting in the model
2. Experiment with prediction threshold
- Lower threshold to catch more "Buy" signals
3. Model tuning
- Adjust hyperparameters
- Try different architectures